tools/migration/internal/runner.go

```diff
@@ -27,22 +40,112 @@ type Migration interface {
 }
 
 type MigrationRunner struct {
-	verbose    bool
-	command    string
+	// verbose controls the amount of output.
+	verbose bool
+
+	// command is the migration command to run, passed from the CLI
+	command string
+
+	// oldRepoOpt is the option passed  from the CLI.
 	oldRepoOpt string
+
+	// MigrationsProvider is a dependency for fetching available migrations
+	// to allow unit tests to supply test migrations without creating test fixtures.
+	MigrationsProvider func() []Migration
+
+	// RepoVersionGetter is a dependency for getting a Repo Version,
+	// to allow the binary version of the repo to be mocked in testing.
+	RepoVersionGetter func() uint
 }
 
 func NewMigrationRunner(verb bool, command, oldRepoOpt string) *MigrationRunner {
-	// TODO: Issue #2585 Implement repo migration version detection and upgrade decisioning
-
 	return &MigrationRunner{
-		verbose:    verb,
-		command:    command,
-		oldRepoOpt: oldRepoOpt,
+		verbose:            verb,
+		command:            command,
+		oldRepoOpt:         oldRepoOpt,
+		MigrationsProvider: DefaultMigrationsProvider,
+		RepoVersionGetter:  DefaultVersionGetter,
 	}
 }
 
+// DefaultVersionGetter fetches the Version constant from the repo package.
+func DefaultVersionGetter() uint {
+	return repo.Version
+}
+
 func (m *MigrationRunner) Run() error {
 	// TODO: Issue #2595 Implement first repo migration
+	repoVersion, err := m.loadVersion()
+	if err != nil {
+		return err
+	}
+	if repoVersion == repo.Version {
+		return errors.New("binary version = repo version; migration not run")
+	}
+	for _, mig := range m.MigrationsProvider() {
+		from, to := mig.Versions()
+		if from == repoVersion && to == from+1 {
```

Comment:
I don't think this is the right form for this check. Clarify that it's just a policy rather than fundamentally necessary to this code that the version bump is exactly one. I can imagine losing this check with no ill effects, if we decide to.

```
if to != from + 1 {
  log.warn("Refusing multi-version migration from %d to %d", ...)
  continue
}
if (from == repoVersion) { ... }